numba/array_analysis.py

```diff
@@ -1152,80 +1203,237 @@ def slice_size(index, dsize):
             size_typ = self.typemap[dsize.name]
             lhs_typ = self.typemap[lhs.name]
             rhs_typ = self.typemap[rhs.name]
-            zero_var = ir.Var(scope, mk_unique_var("zero"), loc)
 
+            if config.DEBUG_ARRAY_OPT >= 2:
+                print("slice_size", "index", index, "dsize", dsize,
+                      "index_def", index_def, "lhs", lhs, "rhs", rhs,
+                      "size_typ", size_typ, "lhs_typ", lhs_typ, "rhs_typ", rhs_typ)
+
+            # Fill in the left side of the slice's ":" with 0 if it wasn't specified.
             if isinstance(lhs_typ, types.NoneType):
+                zero_var = ir.Var(scope, mk_unique_var("zero"), loc)
                 zero = ir.Const(0, loc)
                 stmts.append(ir.Assign(value=zero, target=zero_var, loc=loc))
-                self._define(equiv_set, zero_var, size_typ, zero)
+                self._define(equiv_set, zero_var, types.IntegerLiteral(0), zero)
                 lhs = zero_var
-                lhs_typ = size_typ
+                lhs_typ = types.IntegerLiteral(0)
 
+            # Fill in the right side of the slice's ":" with the array
+            # length if it wasn't specified.
             if isinstance(rhs_typ, types.NoneType):
                 rhs = dsize
                 rhs_typ = size_typ
 
             lhs_rel = equiv_set.get_rel(lhs)
             rhs_rel = equiv_set.get_rel(rhs)
+            if config.DEBUG_ARRAY_OPT >= 2:
+                print("lhs_rel", lhs_rel, "rhs_rel", rhs_rel)
+
+            # Make a deepcopy of the original slice to use as the
+            # replacement slice, which we will modify as necessary
+            # below to convert all negative constants in the slice
+            # to be relative to the dimension size.
+            replacement_slice = copy.deepcopy(index_def)
+            need_replacement = False
+
+            # If the first part of the slice is a constant N then check if N
+            # is negative.  If so, then rewrite the first part of the slice
+            # to be "dsize - N".  This is necessary because later steps will
+            # try to compute slice size with a subtraction which wouldn't work
+            # if any part of the slice was negative.
+            if isinstance(lhs_rel, int):
```

Comment:
Perhaps this branch body should be a function called with `lhs_rel` and `rhs_rel` ? code seems identical apart from addressable side?

Reply:
Some is identical indeed.  Tried to factor that out.